File: typescript/src/debug/place-analysis.ts
/**
 * Analyze places for start/end/environment classification.
 * TypeScript port of Java's PlaceAnalysis.
 */

import type { PetriNet } from '../core/petri-net.js';
import type { Place } from '../core/place.js';

export interface PlaceAnalysisInfo {
  readonly tokenType: string;
  readonly hasIncoming: boolean;
  readonly hasOutgoing: boolean;
}

export class PlaceAnalysis {
  private readonly _data: ReadonlyMap<string, PlaceAnalysisInfo>;

  constructor(data: ReadonlyMap<string, PlaceAnalysisInfo>) {
    this._data = data;
  }

  get data(): ReadonlyMap<string, PlaceAnalysisInfo> {
    return this._data;
  }

  isStart(placeName: string): boolean {
    const info = this._data.get(placeName);
    return info != null && !info.hasIncoming;
  }

  isEnd(placeName: string): boolean {
    const info = this._data.get(placeName);
    return info != null && !info.hasOutgoing;
  }

  /** Build place analysis from a PetriNet. */
  static from(net: PetriNet): PlaceAnalysis {
    const data = new Map<string, { tokenType: string; hasIncoming: boolean; hasOutgoing: boolean }>();

    function ensure(place: Place<unknown>): { tokenType: string; hasIncoming: boolean; hasOutgoing: boolean } {
      let info = data.get(place.name);
      if (!info) {
        info = { tokenType: 'unknown', hasIncoming: false, hasOutgoing: false };
        data.set(place.name, info);
      }
      return info;
    }

    for (const transition of net.transitions) {
      // Input arcs: place → transition (place has outgoing)
      for (const input of transition.inputSpecs) {
        const info = ensure((input as { place: Place<unknown> }).place);
        info.hasOutgoing = true;
      }

      // Output arcs: transition → place (place has incoming)
      if (transition.outputSpec) {
        const outputPlaces = collectOutputPlaces(transition.outputSpec);
        for (const place of outputPlaces) {
          const info = ensure(place);
          info.hasIncoming = true;
        }
      }

      // Inhibitor arcs: just ensure place exists
      for (const inh of transition.inhibitors) {
        ensure((inh as { place: Place<unknown> }).place);
      }

      // Read arcs: place has outgoing (read = test without consuming)
      for (const read of transition.reads) {
        const info = ensure((read as { place: Place<unknown> }).place);
        info.hasOutgoing = true;
      }

      // Reset arcs: just ensure place exists
      for (const reset of transition.resets) {
        ensure((reset as { place: Place<unknown> }).place);
      }
    }

    return new PlaceAnalysis(data as ReadonlyMap<string, PlaceAnalysisInfo>);
  }
}

/** Recursively collect all output places from an Out spec. */
function collectOutputPlaces(out: unknown): Place<unknown>[] {
  const spec = out as { type: string; place?: Place<unknown>; children?: unknown[]; child?: unknown; from?: Place<unknown>; to?: Place<unknown> };
  switch (spec.type) {
    case 'place':
      return spec.place ? [spec.place] : [];
    case 'and':
    case 'xor':
      return (spec.children ?? []).flatMap(c => collectOutputPlaces(c));
    case 'timeout':
      return spec.child ? collectOutputPlaces(spec.child) : [];
    case 'forward-input':
      return spec.to ? [spec.to] : [];
    default:
      return [];
  }
}

File: rust/libpetri-debug/src/place_analysis.rs
//! Analyze places for start/end/environment classification.

use std::collections::HashMap;

use libpetri_core::output;
use libpetri_core::petri_net::PetriNet;

/// Analysis info for a single place.
#[derive(Debug, Clone)]
pub struct PlaceAnalysisInfo {
    pub token_type: String,
    pub has_incoming: bool,
    pub has_outgoing: bool,
}

/// Analyzes all places in a Petri net for structural classification.
#[derive(Debug, Clone)]
pub struct PlaceAnalysis {
    data: HashMap<String, PlaceAnalysisInfo>,
}

impl PlaceAnalysis {
    /// Build place analysis from a PetriNet.
    pub fn from_net(net: &PetriNet) -> Self {
        let mut data: HashMap<String, PlaceAnalysisInfo> = HashMap::new();

        fn ensure<'a>(
            data: &'a mut HashMap<String, PlaceAnalysisInfo>,
            name: &str,
        ) -> &'a mut PlaceAnalysisInfo {
            data.entry(name.to_string())
                .or_insert_with(|| PlaceAnalysisInfo {
                    token_type: "unknown".into(),
                    has_incoming: false,
                    has_outgoing: false,
                })
        }

        for t in net.transitions() {
            // Input arcs: place → transition (place has outgoing)
            for input in t.input_specs() {
                ensure(&mut data, input.place_name()).has_outgoing = true;
            }

            // Output arcs: transition → place (place has incoming)
            if let Some(out) = t.output_spec() {
                for p in output::all_places(out) {
                    ensure(&mut data, p.name()).has_incoming = true;
                }
            }

            // Inhibitor arcs: just ensure place exists
            for inh in t.inhibitors() {
                ensure(&mut data, inh.place.name());
            }

            // Read arcs: place has outgoing
            for r in t.reads() {
                ensure(&mut data, r.place.name()).has_outgoing = true;
            }

            // Reset arcs: just ensure place exists
            for r in t.resets() {
                ensure(&mut data, r.place.name());
            }
        }

        Self { data }
    }

    /// Returns the analysis data.
    pub fn data(&self) -> &HashMap<String, PlaceAnalysisInfo> {
        &self.data
    }

    /// Returns `true` if the place is a start place (no incoming arcs).
    pub fn is_start(&self, place_name: &str) -> bool {
        self.data
            .get(place_name)
            .is_some_and(|info| !info.has_incoming)
    }

    /// Returns `true` if the place is an end place (no outgoing arcs).
    pub fn is_end(&self, place_name: &str) -> bool {
        self.data
            .get(place_name)
            .is_some_and(|info| !info.has_outgoing)
    }
}

#[cfg(test)]
mod tests {
    use super::*;
    use libpetri_core::input::one;
    use libpetri_core::output::out_place;
    use libpetri_core::place::Place;
    use libpetri_core::transition::Transition;

    #[test]
    fn start_and_end_classification() {
        let p1 = Place::<i32>::new("start");
        let p2 = Place::<i32>::new("mid");
        let p3 = Place::<i32>::new("end");

        let t1 = Transition::builder("t1")
            .input(one(&p1))
            .output(out_place(&p2))
            .build();
        let t2 = Transition::builder("t2")
            .input(one(&p2))
            .output(out_place(&p3))
            .build();

        let net = PetriNet::builder("test").transitions([t1, t2]).build();
        let analysis = PlaceAnalysis::from_net(&net);

        assert!(analysis.is_start("start"));
        assert!(!analysis.is_end("start"));
        assert!(!analysis.is_start("mid"));
        assert!(!analysis.is_end("mid"));
        assert!(!analysis.is_start("end"));
        assert!(analysis.is_end("end"));
    }
}
